### ISSUE TITLE
Send events via EventEmitter when a worker is created or destroyed

### ISSUE THREAD

**Issue by pioardi** [pioardi/poolifier#105]
Send events via EventEmitter when a worker is created or destroyed.
This would improve our test design that currently wait a while before to do some checks ( pull mode), we want our tests to be less random and with a push mode.

This could be also be used from end users in some scenarios, for example wait that all the threads are online and ready before to continue execution.

**Comment by Shinigami92** [comment]
We already have a `emitter` in dynamic pool.
There we emit `"FullPool"`

https://github.com/pioardi/poolifier/blob/ba41d390125db582001757a883b0450546e7be2b/src/dynamic.ts#L60

---

I would suggest that we tackle this after #111 and also we should use lowercase events, because (I think) that's more native as other libs (even Node itself) do it this way.
So we would have:
```js
'full'
'created'
'destroyed'
```

**Comment by pioardi** [comment]
Yes we can also change FullPool to fullpool since we are already going to break few little things for retro compatibility

**Comment by Shinigami92** [comment]
I think `'full'` is sufficient, because the class is already a **pool**

**Comment by jerome-benoit** [comment]
How do you plan to tackle the worker exit case? 

If the EventEmitter in that case also emit an event local to the pool, we could then remove it safely from the pool.

**Comment by pioardi** [comment]
> How do you plan to tackle the worker exit case?
> 
> If the EventEmitter in that case also emit an event local to the pool, we could then remove it safely from the pool.

I am not sure I got your point but we can subscribe on the exit event for each worker and then emit it as a pool event

**Comment by pioardi** [comment]
Thinking better about this we already do something like that with onlineHandler and exitHandler, I will try to change our tests based on those already emitted events and also following some tips from [here](https://dev.to/shelob9/testing-event-emitter-38b5)

**Comment by Shinigami92** [comment]
Reopen because it was not solved by #172

**Comment by pioardi** [comment]
> Thinking better about this we already do something like that with onlineHandler and exitHandler, I will try to change our tests based on those already emitted events and also following some tips from [here](https://dev.to/shelob9/testing-event-emitter-38b5)

This is why IMO this is resolved, it is not really needed to emits new events, we already emit events when a worker is created or destroyed.

**Comment by pioardi** [comment]
We could open a new issue if we want to just change the name of the event "FullPool"

**Comment by Shinigami92** [comment]
Oh I see. I thought we want to create additional events like

- `emitter.on('created', ...)`
- `emitter.on('destroyed', ...)`
- `emitter.on('full', ...)`
- `emitter.on('...', ...)`

So a user should always use the handlers instead?
Should we then add a `full` handler?

IMO we could provide both ways

**Comment by pioardi** [comment]
No we don' t want additional events since they can already provide online and exit handlers.
They can already have an error handler, the reason why we do not ask it into the constructors is because is not needed since is not related to the worker units, they can listen it after the pool is created.

:) I don't think we really need to change something, we can just change the "FullPool" to "fullpool" or "full" and we have done

**Comment by Shinigami92** [comment]
I just question it why there are two different ways of doing so?

The handlers are optional callbacks
We could extend the Options for dynamic pools and provide a
```ts
type FullHandler = () => void

// ...

fullHandler?: FullHandler
```

That way, the user only needs to access one way of handling things and we can get rid of the 
internal `class PoolEmitter` at all

But on the other hand, IMO it's a better developer experience if they can use

```ts
myPool.emitter.on('an-event', ...)
// or even better
myPool.on('an-event', ...)
```

This way they can access these events anytime they want and not just while constructing
And in addition: we have a standardized future extensible way of sending events to the developer's pool and they can extend on this behavior

**Comment by Shinigami92** [comment]
But we currently "need" to pass such handlers to the constructor because we pass them to 

https://github.com/pioardi/poolifier/blob/920056d52c7d11ad560fcca9ea3327556188185c/src/pools/abstract-pool.ts#L315-L317

:thinking: 

---

Mh so I see: the passed options handlers are listeners for the workers itself and not the pool :thinking:

**Comment by Shinigami92** [comment]
IMO we could make the emitter protected and then define `on` events as public member functions for wrapping the emitter calls
Maybe `AbstractPool extends EventEmitter` also could work, but then we would have everything of the EventEmitter :confused:

**Comment by pioardi** [comment]
> But we currently "need" to pass such handlers to the constructor because we pass them to
> 
> https://github.com/pioardi/poolifier/blob/920056d52c7d11ad560fcca9ea3327556188185c/src/pools/abstract-pool.ts#L315-L317
> 
> 🤔
> 
> Mh so I see: the passed options handlers are listeners for the workers itself and not the pool 🤔

Correct option handlers are to listen worker events.

**Comment by pioardi** [comment]
> IMO we could make the emitter protected and then define `on` events as `public` member functions for wrapping the emitter calls
> Maybe `AbstractPool extends EventEmitter` also could work, but then we would have everything of the EventEmitter 😕

We should close this issue and start a discussion on it if you want.
I personally don't see big benefits to change the current behavior.

**Comment by Shinigami92** [comment]
Yeah, IMO it could be a change for a later version anyways
